### PR TITLE
Start migrating services.UserService to use context and return users

### PR DESCRIPTION
**Pull request by rosstimothy**
Adds new variants of existing methods that are going to be updated to support propagating context and return users from create, update and upsert. This is an unfortunate step required because e utilizes the interface for various functionality. In order to prevent breaking builds, the temporary methods were added so that e can be converted to them first, then oss can be updated to the new version of the interface. Once that is done e will be converted and then the temp methods will be removed.

Implements step 1 of https://github.com/gravitational/teleport/issues/32949

### CODE REVIEW

File: api/client/client.go [hunk 1]
```diff
@@ -838,6 +838,18 @@ func (c *Client) CreateUser(ctx context.Context, user types.User) error {
...
+// TODO(tross) remove this once oss and e are converted to using the new signature.
+func (c *Client) CreateUserWithContext(ctx context.Context, user types.User) (types.User, error) {
+	err := c.CreateUser(ctx, user)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	created, err := c.GetUserWithContext(ctx, user.GetName(), false)
+	return created, trace.Wrap(err)
+}
```
Comment (espadolini) [inline api/client/client.go:851]: What happens if someone deletes the user between a successful create and the get?

What happens if the user has create permissions but not read permissions?

File: lib/services/local/users.go [hunk 2]
```diff
@@ -287,10 +328,11 @@ func (s *IdentityService) CompareAndSwapUser(ctx context.Context, new, existing
...
 	existingItem := backend.Item{
-		Key:     backend.Key(webPrefix, usersPrefix, existing.GetName(), paramsPrefix),
-		Value:   existingValue,
-		Expires: existing.Expiry(),
-		ID:      existing.GetResourceID(),
+		Key:      backend.Key(webPrefix, usersPrefix, existing.GetName(), paramsPrefix),
+		Value:    existingValue,
+		Expires:  existing.Expiry(),
+		ID:       existing.GetResourceID(),
+		Revision: existing.GetRevision(),
```
Comment (espadolini) [inline lib/services/local/users.go:335]: I can understand wanting to be consistent in the other functions (although we only need to pass along revision and ID on upsert, for the sake of the cache) but for the existing item on a CAS there's really no reason.
```suggestion
```

File: lib/services/local/users.go [hunk 3]
```diff
@@ -148,49 +168,60 @@ func (s *IdentityService) CreateUser(user types.User) error {
...
 	}
 	if auth := user.GetLocalAuth(); auth != nil {
-		if err = s.upsertLocalAuthSecrets(user.GetName(), *auth); err != nil {
-			return trace.Wrap(err)
+		if err = s.upsertLocalAuthSecrets(ctx, user.GetName(), *auth); err != nil {
+			return nil, trace.Wrap(err)
 		}
 	}
-	return nil
+	user.SetRevision(lease.Revision)
```
Comment (espadolini) [inline lib/services/local/users.go:223]: Should we zero the ID?

File: lib/auth/api.go [hunk 4]
```diff
@@ -779,7 +797,7 @@ type ReadOktaAccessPoint interface {
 	// ListOktaAssignments returns a paginated list of all Okta assignment resources.
 	ListOktaAssignments(context.Context, int, string) ([]types.OktaAssignment, string, error)
 
-	// GetOktaAssignmen treturns the specified Okta assignment resources.
+	// GetOktaAssignment treturns the specified Okta assignment resources.
```
Comment (espadolini) [inline lib/auth/api.go:800]:
```suggestion
	// GetOktaAssignment returns the specified Okta assignment resource.
```

File: lib/services/local/users.go (hunk 3 above)
Comment (espadolini) [reply lib/services/local/users.go:223]: Probably not, but of all the possible values, 0 is less offensive than "whatever was in the resource that the caller passed in".

Comment (espadolini) [reply lib/services/local/users.go:223]: In `Get` we set the resource ID at unmarshal time, from the actual ID returned by the backend. We don't have the ID from the lease here, so we either leave whatever random ID was in the resource passed by the caller or we set it to 0, and I think that the latter is preferrable. We could also set some other sentinel value like `0x7fff_ffff_ffff_ffff` if you prefer.

**Review by espadolini — APPROVED** [review]

**Review by zmb3 — APPROVED** [review]

**Pull request merged**